### DISCUSSION TITLE
edge ordering in undirected graph

### DISCUSSION THREAD

**Discussion by errhernandez** [pyg-team/pytorch_geometric#3020] [Q&A] [answered]
In an undirected graph each edge has to be specified twice; presumably the order of edges in the edge_index array is irrelevant, but the question is: is there a preferred order in terms of efficiency? For example, take the graph illustration in the "Introduction by example" doc, where:
edge_index = torch.tensor([[0, 1, 1, 2],
                           [1, 0, 2, 1]], dtype=torch.long)
Presumably the following would be equally valid:
edge_index = torch.tensor([[0, 1, 1, 2],
                           [1, 2, 0, 1]], dtype=torch.long)
In other words, my question is: Is it desirable (in terms of efficiency of execution) that the two directions of an undirected edge are specified contiguously in edge_index?
Thanks.

**Comment by rusty1s** [comment, marked as answer]
There are only minor differences in terms of efficiency related to the ordering of edges, and it's not really controllable by the user. The efficiency solely depends on how many atomic operations overlap within parallel CUDA threads that point to the same target node and same feature index. As such, source/row-wise ordering or random ordering will likely perform equal, while target/col-wise ordering can bot does not have be slower.
